L12-csharp-interceptors.md

```diff
@@ -0,0 +1,212 @@
+underlying `CallInvoker` handler for the final interceptor in the chain:
+
+```csharp
+TResponse BlockingUnaryCall<TRequest, TResponse>(Method<TRequest, TResponse> method, string host, CallOptions options, TRequest request, Func<Method<TRequest, TResponse>, string, CallOptions, TRequest, TResponse> next);
+AsyncUnaryCall<TResponse> AsyncUnaryCall<TRequest, TResponse>(Method<TRequest, TResponse> method, string host, CallOptions options, TRequest request, Func<Method<TRequest, TResponse>, string, CallOptions, TRequest, AsyncUnaryCall<TResponse>> next);
+AsyncServerStreamingCall<TResponse> AsyncServerStreamingCall<TRequest, TResponse>(Method<TRequest, TResponse> method, string host, CallOptions options, TRequest request, Func<Method<TRequest, TResponse>, string, CallOptions, TRequest, AsyncServerStreamingCall<TResponse>> next);
+AsyncClientStreamingCall<TRequest, TResponse> AsyncClientStreamingCall<TRequest, TResponse>(Method<TRequest, TResponse> method, string host, CallOptions options, Func<Method<TRequest, TResponse>, string, CallOptions, AsyncClientStreamingCall<TRequest, TResponse>> next);
+AsyncDuplexStreamingCall<TRequest, TResponse> AsyncDuplexStreamingCall<TRequest, TResponse>(Method<TRequest, TResponse> method, string host, CallOptions options, Func<Method<TRequest, TResponse>, string, CallOptions, AsyncDuplexStreamingCall<TRequest, TResponse>> next);
+```
+
+In the general case, the interceptors are not obligated to invoke `next` once.
+They might choose to not continue with the call chain and terminate it
+immediately by returning their own desired return value, or potentially call
+`next` more than once, to simulate retries, transactions, or other similar
+potential use cases.  In fact, the ability to take an explicit `next` argument
+is the primary rationale for needing a special-purpose interceptor machinery on
+the client side, since `CallInvoker` is capable of doing a lot for us.
+However, chaining `CallInvoker`s require each one to explicitly hold a
+reference to the next `CallInvoker` in the chain.  The `ClientInterceptor`
+design, however, abstracts away the next function from the object and passes it
+as a continuation to each interceptor hook.  An internal `CallInvoker`
+implementation chains interceptors together and ultimately with the underlying
+`CallInvoker` and `Channel` objects.  All of the functionality described so far
+could have been implemented as an external library, and does not change the
+non-intercepted code path at all.  In addition to hook functions to be
+overridden by the client interceptor implementations, the class provides a few
+`static` methods that lets the interceptor register additional hooks to execute
+code when certain events happen in asynchronous and streaming calls.  These
+helper static functions are the only reason this facility could not have been
+implemented by an external library: they require access to the internal
+constructors of `AsyncUnaryCall<TResponse>`,
```

Comment:
FYI we've actually been thinking of making those constructors public (they're useful for testing as well).